File: 백준/Gold/2110. 공유기 설치/공유기 설치.cc
#include <bits/stdc++.h>

using namespace std;

int main(int argc, char *argv[]) {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int n, c;
  cin >> n >> c;
  vector<int> arr(n);
  for (int &x : arr) {
    cin >> x;
  }

  sort(arr.begin(), arr.end());
  int lo = 1, hi = arr.back() - arr[0];

  int ans = 0;

  while (lo <= hi) {
    int mid = (lo + hi) / 2;
    int last = arr[0];
    int cnt = 1;
    bool flag = false;
    for (int i = 1; i < n; ++i) {
      if (arr[i] - last >= mid) {
        last = arr[i];
        cnt++;
        if (cnt == c) {
          flag = true;
          break;
        }
      }
    }

    if (flag) {
      ans = mid;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  cout << ans << "\n";

  return 0;
}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, c int32
	fmt.Fscan(reader, &n, &c)
	arr := make([]int32, n)
	for i := int32(0); i < n; i++ {
		fmt.Fscan(reader, &arr[i])
	}
	sort.Slice(arr, func(i, j int) bool {
		return arr[i] < arr[j]
	})

	lo := int32(1)
	hi := int32(arr[n-1] - arr[0])
	for lo <= hi {
		mid := (lo + hi) / 2
		flag := false
		last := int32(arr[0])
		cnt := int32(1)
		for i := int32(1); i < n; i++ {
			if arr[i]-last >= mid {
				last = arr[i]
				cnt++
				if cnt == c {
					flag = true
					break
				}
			}
		}
		if flag {
			lo = mid + 1
		} else {
			hi = mid - 1
		}
	}

	fmt.Fprintln(writer, lo-1)
}